lib/active_storage/service/mirror_service.rb

```diff
@@ -18,24 +18,24 @@ def initialize(primary:, mirrors:)
   end
 
   def upload(key, io, checksum: nil)
-    each_service.collect do |service|
-      service.upload key, io.tap(&:rewind), checksum: checksum
+    perform_async_across_services do |service|
+      service.upload key, StringIO.new(io.tap(&:rewind).read), checksum: checksum
```

Comment:
Right. I said we can’t just share a single IO instance between multiple threads, so cloning the given IO instance for each thread was a sensible thing to try. But we can’t do that, either: each clone will share the same underlying file descriptor.

So what can we do? Jeremy pointed you to [ruby-tee](https://github.com/lloeki/ruby-tee/blob/master/lib/tee.rb). We probably can’t use it as-is, but its implementation is a good example of a way to do exactly what we want to do: safely share a single IO between multiple threads.